File: py/test_money.py
from shutil import ExecError
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
import unittest
from money import Money
from portfolio import Portfolio
from bank import Bank


class TestMoney(unittest.TestCase):

    def setUp(self) -> None:
        self.bank = Bank()
        self.bank.addExchangeRate("EUR", "USD", 1.2)
        self.bank.addExchangeRate("USD", "KRW", 1100)

    def testMultiplication(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = Money(20, "EUR")
        self.assertEqual(tenEuros.times(2), twentyEuros)

    def testDivision(self):
        originalMoney = Money(4002, "KRW")
        actualMoneyAfterDivision = originalMoney.divide(4)
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        self.assertEqual(expectedMoneyAfterDivision, actualMoneyAfterDivision)

    def testAddition(self):
        fiveDollars = Money(5, "USD")
        tenDollars = Money(10, "USD")
        fifteenDollars = Money(15, "USD")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfolio.evaluate(self.bank, "USD"))

    def testAddingEurAndUsd(self):
        fiveDollars = Money(5, "USD")
        tenEuros = Money(10, "EUR")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenEuros)
        expected = Money(17, "USD")
        actual = portfolio.evaluate(self.bank, "USD")
        self.assertEqual(actual, expected,
                         "Expected %s, but got %s" % (expected, actual))

    # 1 USD + 1100 KRW = 2200 KRW
    def testAddingUsdAndWon(self):
        oneDollar = Money(1, "USD")
        elevenHundredWons = Money(1100, "KRW")
        portfolio = Portfolio()
        portfolio.add(oneDollar, elevenHundredWons)
        expected = Money(2200, "KRW")
        actual = portfolio.evaluate(self.bank, "KRW")
        self.assertEqual(actual, expected,
                         "Expected %s, but got %s" % (expected, actual))

    # handle unknown exchange rates
    def testAdditionOfMultipleMissingExchangeRates(self):
        oneDollar = Money(1, "USD")
        oneEuro = Money(1, "EUR")
        oneWon = Money(1, "KRW")
        portfolio = Portfolio()
        portfolio.add(oneDollar, oneEuro, oneWon)
        with self.assertRaisesRegex(
            Exception,
            r"Missing exchange rate\(s\):\[USD\->Kalganid,EUR\->Kalganid,KRW\->Kalganid\]",
        ):
            portfolio.evaluate(self.bank, "Kalganid")

    def testConversionWitDifferentRatesBetweenTwoCurrencies(self):
        tenEuros = Money(10, "EUR")
        actual = self.bank.convert(tenEuros, "USD")
        self.assertEqual(actual, Money(12, "USD"))

        self.bank.addExchangeRate("EUR", "USD", 1.3)
        actual = self.bank.convert(tenEuros, "USD")
        self.assertEqual(actual, Money(13, "USD"))

    def testConversionWithMissingExchangeRates(self):
        tenEuros = Money(10, "EUR")
        with self.assertRaisesRegex(
            Exception,
            r"EUR\->Kalganid"
        ):
            self.bank.convert(tenEuros, "Kalganid")

    # allow exchange rates to be modified


if __name__ == '__main__':
    unittest.main()

File: js/test_money.js
const { AssertionError } = require("assert");
const assert = require("assert");
const Money = require("./money");
const Portfolio = require("./portfolio");
const Bank = require("./bank");

class MoneyTests {

    constructor() {
        this.bank = new Bank();
        this.bank.addExchangeRate("EUR", "USD", 1.2);
        this.bank.addExchangeRate("USD", "KRW", 1100);
    }

    testMultiplication() {
        let tenEuros = new Money(10, "EUR");

        let actual = tenEuros.times(2);

        let expected = new Money(20, "EUR");
        assert.deepStrictEqual(actual, expected);
    }

    testDivision() {
        let originalMoney = new Money(4002, "KRW");

        let actual = originalMoney.divide(4);

        let expected = new Money(1000.5, "KRW");
        assert.deepStrictEqual(actual, expected);
    }

    testAddition() {
        let fiveDollars = new Money(5, "USD");
        let tenDollars = new Money(10, "USD");
        let portfolio = new Portfolio();
        
        portfolio.add(fiveDollars, tenDollars);
        
        const expected = new Money(15, "USD");
        const actual = portfolio.evaluate(this.bank, "USD");
        assert.deepStrictEqual(actual, expected);
    }

    testAdditionOfEurosAndDollars() {
        let fiveDollars = new Money(5, "USD");
        let tenEuros = new Money(10, "EUR");
        let portfolio = new Portfolio();

        portfolio.add(fiveDollars, tenEuros);

        const expected = new Money(17, "USD");
        const actual = portfolio.evaluate(this.bank, "USD");
        assert.deepStrictEqual(actual, expected);
    }

    // 1 USD + 1100 KRW = 2200 KRW
    testAdditionOfDollarsAndWons() {
        let oneDollar = new Money(1, "USD");
        let elevenHundredWon = new Money(1100, "KRW");
        let portfolio = new Portfolio();

        portfolio.add(oneDollar, elevenHundredWon);

        const expected = new Money(2200, "KRW");
        const actual = portfolio.evaluate(this.bank, "KRW");
        assert.deepStrictEqual(actual, expected);
    }

    // handle unknown exchange rates
    testAdditionOfMultipleMissingExchangeRates() {
        let oneDollar = new Money(1, "USD");
        let oneEuro = new Money(1, "EUR");
        let oneWon = new Money(1, "KRW");
        let portfolio = new Portfolio();

        portfolio.add(oneDollar, oneEuro, oneWon);

        let expectedError = "Missing exchange rate(s): [USD->Kalganid,EUR->Kalganid,KRW->Kalganid]"
        assert.throws(() => portfolio.evaluate("Kalganid"), expectedError);
    }

    // refactor conversion key generator
    testConversion() {
        const tenEuros = new Money(10, "EUR");

        let actual = this.bank.convert(tenEuros, "USD");

        assert.deepStrictEqual(actual, new Money(12, "USD"));
    }

    testConversionWithMissingExchangeRates() {
        const tenEuros = new Money(10, "EUR");

        const expectedError = new Error("EUR->Kalganid");
        assert.throws(() => this.bank.convert(tenEuros, "Kalganid"), expectedError);
    }



    // allow exchange rates to be modified



    getAllTestMethods() {
        let proto = Object.getPrototypeOf(this);
        let members = Object.getOwnPropertyNames(proto);
        let testNames = members.filter(p => { return typeof proto[p] === "function" && p.startsWith("test") });
        return testNames;
    }

    runAllTests() {
        console.log();

        let tests = this.getAllTestMethods();
        console.log("Discovered %d test(s)", tests.length);

        let i = 0, failures = 0;
        tests.forEach((testName) => {
            let test = Reflect.get(this, testName);
            console.log("Running %d/%d: %s()", ++i, tests.length, testName);
            try {
                Reflect.apply(test, this, []);
            }
            catch (e) {
                failures++;
                if (e instanceof AssertionError) {
                    this.print(e, "red");
                }
                else {
                    this.print("Unhandled error", "red");
                    throw (e);
                }
            }
        });

        console.log();
        if (failures) {
            this.print(`${failures} test(s) failed`, "red")
        }
        this.print(`${i - failures} test(s) succeeded`, "green");
        console.log();

    }

    print(msg, color) {
        if (!color) {
            console.log(msg);
        }
        else {
            let hexColor = ""
            if (color === "red") {
                hexColor = "\x1b[31m"
            }
            else if (color === "green") {
                hexColor = "\x1b[32m";
            }

            console.log('%s%s\x1b[0m', hexColor, msg);
        }
    }
}



new MoneyTests().runAllTests();